contracts/Unipool.sol

```diff
@@ -135,8 +154,102 @@ contract Unipool is LPTokenWrapper {
         lastUpdateTime = block.timestamp;
         periodFinish = block.timestamp.add(DURATION);
 
-        tradedToken.safeTransferFrom(msg.sender, address(this), _amount);
+        rewardToken.safeTransferFrom(msg.sender, address(this), _amount);
 
         emit RewardAdded(_amount);
     }
+    
+    function reinvestReward() public updateReward(msg.sender) onlyTradable() {
+        uint256 reward = earned(msg.sender);
+        require(reward > 0, "Nothing to reinvest");
+        
+        (uint256 liquidity, uint256 remainingReward) = _poolReward(reward);
+        rewards[msg.sender] = remainingReward;
+        super.stakeInternal(liquidity);
+
+        emit ReinvestedReward(msg.sender, reward.sub(remainingReward), liquidity);
+    }
+
+    function _getMinimumTokensForRewardSale(uint256 tradeSize) private view returns(uint256 minTokens) {
+        IUniswapV2Pair pair = IUniswapV2Pair(address(uniswapTokenExchange));
+
+        // Get reserves and order them correctly
+        (uint256 reserve0, uint256 reserve1,) = pair.getReserves();
+        (uint256 rewardReserve, uint256 foreignReserve) = 
+            rewardToken == pair.token0() ? (reserve0, reserve1) : (reserve1, reserve0);
+        
+        // Price of our trade if the price didn't slip
+        uint256 immediatePrice = tradeSize.mul(foreignReserve).div(rewardReserve);
+        // The lowest we'd be willing to accept, based on max slippage
+        uint256 acceptablePrice = immediatePrice.mul(10000 - MAX_SLIPPAGE_BASIS_POINTS).div(10000);
+        // Preflight here mostly for the benefit of unit tests
+        require(uniswapRouter.getAmountOut(tradeSize, rewardReserve, foreignReserve) >= acceptablePrice, 
+            "Slippage too high for automagic reinvestment");
+        return acceptablePrice;
+    }
+    
+    function _poolReward(uint256 rewardAmount) private returns(uint256 liquidityTokens, uint256 remainingReward) {
+        // convert half to target token
+        uint256 rewardToConvert = rewardAmount.div(2);
```

Comment:
You can actually just divide by 2 here. The only check `div()` does is that you don't divide by 0.